src/ecdsa/ellipticcurve.py

```diff
@@ -1324,8 +1327,41 @@ def from_bytes(
             curve, data, validate_encoding, valid_encodings
         )
         return PointEdwards(
-            curve, coord_x, coord_y, 1, coord_x * coord_y, order
+            curve, coord_x, coord_y, 1, coord_x * coord_y, order, generator
+        )
+
+    def _maybe_precompute(self):
+        if not self.__generator or self.__precompute:
+            return
+
+        # since this code will execute just once, and it's fully deterministic,
+        # depend on atomicity of the last assignment to switch from empty
+        # self.__precompute to filled one and just ignore the unlikely
+        # situation when two threads execute it at the same time (as it won't
+        # lead to inconsistent __precompute)
+        order = self.__order
+        assert order
+        precompute = []
+        i = 1
+        order *= 2
+        coord_x, coord_y, coord_z, coord_t = self.__coords
+        prime = self.__curve.p()
+        doubler = PointEdwards(
+            self.__curve, coord_x, coord_y, coord_z, coord_t, order
         )
+        order *= 2
+        coord_x, coord_y = doubler.x(), doubler.y()
+        coord_t = coord_x * coord_y % prime
+        precompute.append((coord_x, coord_y, coord_t))
+
+        while i < order:
+            i *= 2
+            doubler = doubler.double().scale()
+            coord_x, coord_y = doubler.x(), doubler.y()
+            coord_t = coord_x * coord_y % prime
+            precompute.append((coord_x, coord_y, coord_t))
+
+        self.__precompute = precompute
```

Comment:
Here are two attempts to restructure the do-while awkwardness, with control flow

``` python
        assert self.__order
        precompute = []
        prime = self.__curve.p()
        doubler = PointEdwards(self.__curve, *self.__coords, self.__order * 2)
        i = 1
        while True:
            doubler.scale()
            coord_x, coord_y = doubler.x(), doubler.y()
            coord_t = coord_x * coord_y % prime
            precompute.append((coord_x, coord_y, coord_t))
            if i >= self.__order * 4:
                break
            i *= 2
            doubler = doubler.double()
        self.__precompute = precompute
        # assert self.__order * 8 < 2**len(precompute) < self.__order * 16
```

and with extracting scale-yield-double to a separate generator

``` python
        def double_gen(p):
            while True:
                yield p.scale()
                p = p.double()

        assert self.__order
        prime = self.__curve.p()
        p1 = PointEdwards(self.__curve, *self.__coords, self.__order * 2)
        import itertools
        self.__precompute = list(itertools.islice(
            ((p.x(), p.y(), p.x() * p.y() % prime) for p in double_gen(p1)),
            bit_length(self.__order) + 3
        ))
        # assert self.__order * 8 < 2**len(self.__precompute) < self.__order * 16
```

Reply:
yes, `*self.__coords` is nice syntactic sugar, unfortunately doesn't work on python2, but yeah, could be rewritten